examples/computecloth/computecloth.cpp

```diff
@@ -595,14 +651,35 @@ class VulkanExample : public VulkanExampleBase
 		VK_CHECK_RESULT(vkCreateCommandPool(device, &cmdPoolInfo, nullptr, &compute.commandPool));
 
 		// Create a command buffer for compute operations
-		VkCommandBufferAllocateInfo cmdBufAllocateInfo = vks::initializers::commandBufferAllocateInfo(compute.commandPool, VK_COMMAND_BUFFER_LEVEL_PRIMARY, 2);
+		VkCommandBufferAllocateInfo cmdBufAllocateInfo = vks::initializers::commandBufferAllocateInfo(compute.commandPool, VK_COMMAND_BUFFER_LEVEL_PRIMARY, 1);
 		VK_CHECK_RESULT(vkAllocateCommandBuffers(device, &cmdBufAllocateInfo, &compute.commandBuffers[0]));
 
 		// Semaphores for graphics / compute synchronization
 		VkSemaphoreCreateInfo semaphoreCreateInfo = vks::initializers::semaphoreCreateInfo();
 		VK_CHECK_RESULT(vkCreateSemaphore(device, &semaphoreCreateInfo, nullptr, &compute.semaphores.ready));
 		VK_CHECK_RESULT(vkCreateSemaphore(device, &semaphoreCreateInfo, nullptr, &compute.semaphores.complete));
 
+		// Acquire the input buffer from the graphics queue explicitly here.  This is the only time we need to do this
+		// as the input buffer doesn't move between queues after initialization.  Note, this wouldn't be necessary
+		// if we created the input buffer on the compute queue in the first place, but the order of initialization
+		// means both storage buffers are created on the graphics queue.
+		// There are alternative paths, but all would be complicated by the fact that we need the input buffer
+		// populated with the initial data, and the output buffer to be initially owned by the graphics queue with a
+		// pending release barrier to the compute queue, to line up with the first compute dispatch.
+		VkCommandBuffer acquireCmd = vulkanDevice->createCommandBuffer(VK_COMMAND_BUFFER_LEVEL_PRIMARY, compute.commandPool, true);
+
+		VkBufferMemoryBarrier bufferBarrier = vks::initializers::bufferMemoryBarrier();
+		bufferBarrier.srcAccessMask = 0;
+		bufferBarrier.dstAccessMask = VK_ACCESS_SHADER_READ_BIT;
+		bufferBarrier.srcQueueFamilyIndex = vulkanDevice->queueFamilyIndices.graphics;
+		bufferBarrier.dstQueueFamilyIndex = vulkanDevice->queueFamilyIndices.compute;
+		bufferBarrier.size = VK_WHOLE_SIZE;
+		bufferBarrier.buffer = storageBuffers.input.buffer;
+		// Explicitly acquire the storageBuffer.input buffer (the output buffer will ping-pong back and forth so it's acquired as part of the main compute command buffer)
+		vkCmdPipelineBarrier(acquireCmd, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, VK_FLAGS_NONE, 0, nullptr, 1, &bufferBarrier, 0, nullptr);
```

Comment:
this queue transition could be eliminated if we deferred populating the input buffer with the initial contents until here and did it explicitly on the compute queue.  That would probably be cleaner since we'd no longer ever need to move the input buffer across queues, but we'd still need to explicitly have the graphics queue take ownership of the output buffer and release it to the compute queue.

Can you have a queue take ownership by creating a command buffer that does _nothing_ but execute a release pipeline barrier?

Reply:
I was initially avoiding this because I thought the resulting diff would look much larger because of moved code, but realized I can avoid moves by factoring out the particle vector computation to its own function.

Reply:
OK, the diff is longer than I'd like, but the input buffer only ever gets used on the compute queue now.